src/gui/torrentcategorydialog.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2017  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "torrentcategorydialog.h"
+
+#include <QMap>
+#include <QMessageBox>
+
+#include "base/bittorrent/session.h"
+#include "ui_torrentcategorydialog.h"
+
+TorrentCategoryDialog::TorrentCategoryDialog(QWidget *parent)
+    : QDialog {parent}
+    , m_ui {new Ui::TorrentCategoryDialog}
+{
+    m_ui->setupUi(this);
+    m_ui->comboSavePath->setMode(FileSystemPathEdit::Mode::DirectorySave);
+    m_ui->comboSavePath->setDialogCaption(tr("Choose save path"));
+}
+
+TorrentCategoryDialog::~TorrentCategoryDialog()
+{
+    delete m_ui;
+}
+
+QString TorrentCategoryDialog::createCategory(QWidget *parent, const QString &parentCategoryName)
+{
+    using BitTorrent::Session;
+
+    QString newCategoryName {parentCategoryName.trimmed()};
+    if (!newCategoryName.isEmpty())
+        newCategoryName += QLatin1Char('/');
+    newCategoryName += tr("New Category");
+
+    TorrentCategoryDialog dialog(parent);
+    dialog.setCategoryName(newCategoryName);
+    while (dialog.exec() == TorrentCategoryDialog::Accepted) {
+        newCategoryName = dialog.categoryName();
+
+        if (!BitTorrent::Session::isValidCategoryName(newCategoryName)) {
+            QMessageBox::warning(
+                        parent, tr("Invalid category name")
+                        , tr("Category name must not contain '\\'.\n"
```

Comment:
'Category' is an object her (not a subject), hence it can't decide for itself, and therefore 'may not' seems to be more appropriate here. @LordNyriox, please?

Reply:
'may not' instead of 'must'???
I am not a good connoisseur of English, but in my opinion, is not suitable here.
It MAY or MAY NOT contain these characters? No. There's no choice! It strictly MUST NOT contain these characters.

Reply:
It seems right to me. Maybe add a "The" in the beginning of the sentence.